packages/sdk38/src/lcdclient.ts

```diff
@@ -0,0 +1,350 @@
+    A extends LcdModule,
+    B extends LcdModule,
+    C extends LcdModule,
+    D extends LcdModule,
+    E extends LcdModule
+  >(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+    setupModuleB: LcdModuleSetup<B>,
+    setupModuleC: LcdModuleSetup<C>,
+    setupModuleD: LcdModuleSetup<D>,
+    setupModuleE: LcdModuleSetup<E>,
+  ): LcdClient & A & B & C & D & E;
+
+  /** Constructs an LCD client with 6 modules */
+  public static withModules<
+    A extends LcdModule,
+    B extends LcdModule,
+    C extends LcdModule,
+    D extends LcdModule,
+    E extends LcdModule,
+    F extends LcdModule
+  >(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+    setupModuleB: LcdModuleSetup<B>,
+    setupModuleC: LcdModuleSetup<C>,
+    setupModuleD: LcdModuleSetup<D>,
+    setupModuleE: LcdModuleSetup<E>,
+    setupModuleF: LcdModuleSetup<F>,
+  ): LcdClient & A & B & C & D & E & F;
+
+  /** Constructs an LCD client with 7 modules */
+  public static withModules<
+    A extends LcdModule,
+    B extends LcdModule,
+    C extends LcdModule,
+    D extends LcdModule,
+    E extends LcdModule,
+    F extends LcdModule,
+    G extends LcdModule
+  >(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+    setupModuleB: LcdModuleSetup<B>,
+    setupModuleC: LcdModuleSetup<C>,
+    setupModuleD: LcdModuleSetup<D>,
+    setupModuleE: LcdModuleSetup<E>,
+    setupModuleF: LcdModuleSetup<F>,
+    setupModuleG: LcdModuleSetup<G>,
+  ): LcdClient & A & B & C & D & E & F & G;
+
+  /** Constructs an LCD client with 8 modules */
+  public static withModules<
+    A extends LcdModule,
+    B extends LcdModule,
+    C extends LcdModule,
+    D extends LcdModule,
+    E extends LcdModule,
+    F extends LcdModule,
+    G extends LcdModule,
+    H extends LcdModule
+  >(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+    setupModuleB: LcdModuleSetup<B>,
+    setupModuleC: LcdModuleSetup<C>,
+    setupModuleD: LcdModuleSetup<D>,
+    setupModuleE: LcdModuleSetup<E>,
+    setupModuleF: LcdModuleSetup<F>,
+    setupModuleG: LcdModuleSetup<G>,
+    setupModuleH: LcdModuleSetup<H>,
+  ): LcdClient & A & B & C & D & E & F & G & H;
+
+  public static withModules<
+    A extends LcdModule,
+    B extends LcdModule,
+    C extends LcdModule,
+    D extends LcdModule,
+    E extends LcdModule,
+    F extends LcdModule,
+    G extends LcdModule,
+    H extends LcdModule
+  >(
+    options: LcdClientBaseOptions,
+    setupModuleA?: LcdModuleSetup<A>,
+    setupModuleB?: LcdModuleSetup<B>,
+    setupModuleC?: LcdModuleSetup<C>,
+    setupModuleD?: LcdModuleSetup<D>,
+    setupModuleE?: LcdModuleSetup<E>,
+    setupModuleF?: LcdModuleSetup<F>,
+    setupModuleG?: LcdModuleSetup<G>,
+    setupModuleH?: LcdModuleSetup<H>,
+  ): any {
+    const client = new LcdClient(options.apiUrl, options.broadcastMode);
+
+    const modules = new Array<LcdModule>();
```

Comment:
Huh, wouldn't it be simpler to just take the `Array<LcdModule>`? I mean maybe this is nicer to caller, but why so complex?

Reply:
Overloading with different generic types for each argument is required for type safety as mentioned above.